sdpa_op.py

```diff
@@ -0,0 +1,177 @@
+    reapply_views = torch._C._functionalization_reapply_views_tls()
+
+    q, k, v, *other_buffers = _unwrap_all_tensors_from_functional((q, k, v, *other_buffers), reapply_views=reapply_views)
+    with _ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.Functionalize)):
+        out = sdpa(q, k, v, score_mod, *other_buffers)
+        return _wrap_all_tensors_to_functional(out, level=0)
+
+@sdpa.py_impl(FakeTensorMode)
+def sdpa_fake_tensor_mode(*args, **kwargs):
+    return sdpa_dense(*args, **kwargs)
+
+sdpa.fallthrough(DispatchKey.PythonDispatcher)
+sdpa.fallthrough(DispatchKey.PythonTLSSnapshot)
+sdpa.fallthrough(DispatchKey.ADInplaceOrView)
+sdpa.fallthrough(DispatchKey.BackendSelect)
+sdpa.fallthrough(DispatchKey.AutocastCPU)
+sdpa.fallthrough(DispatchKey.AutocastCPU)
+
+def bench(f, name=None, iters=100, warmup=5, display=True, profile=False):
+    import time
+    from triton.testing import do_bench
+
+    for _ in range(warmup):
+        f()
+    if profile:
+        with torch.profiler.profile() as prof:
+            f()
+        prof.export_chrome_trace(f"{name if name is not None else 'trace'}.json")
+
+    us_per_iter = do_bench(lambda: f())*1000
+
+    if name is None:
+        res = us_per_iter
+    else:
+        res= f"{name}: {us_per_iter:.3f}us"
+
+    if display:
+        print(res)
+    return res
+
+import torch
+
+Z = 4
+H = 8
+N_CTX = 2048
+D_HEAD = 64
+dtype = torch.float16
+torch.manual_seed(0)
+q = torch.randn((Z, H, N_CTX, D_HEAD), dtype=dtype, device="cuda")
+k = torch.randn((Z, H, N_CTX, D_HEAD), dtype=dtype, device="cuda")
+v = torch.randn((Z, H, N_CTX, D_HEAD), dtype=dtype, device="cuda")
+torch.set_default_device('cuda')
+
+vals = torch.randn(N_CTX, dtype=dtype)
+def identity(score, b, h, m, n):
+    return score
+
+def causal_mask(score, b, h, m, n):
+    return torch.where(m <= n, score, float("-inf"))
+
+def rel_bias(score, b, h, m, n):
+    bias = (m - n)
+    return score + bias
+
+def compose(*fs):
+    def new_func(score, b, h, m, n):
+        for f in fs:
+            score = f(score, b, h, m, n)
+        return score
+    return new_func
+
+def alibi_bias(score, b, h, m, n):
+    bias = (m - n) * h
+    return score + bias
+
+def create_attention(score_mod):
+    def foo(q, k, v):
+        return sdpa(q, k, v, score_mod)
+    return foo
+
+score_mods = {
+    "nop": identity,
+    "causal": causal_mask,
+    "rel": rel_bias,
+    "alibi": alibi_bias,
+    "rel + causal": compose(rel_bias, causal_mask),
+    "alibi + causal": compose(alibi_bias, causal_mask),
+    }
+
+for name, score_mod in score_mods.items():
+    foo = create_attention(score_mod)
+    compiled = torch.compile(foo)
+
+    ref_out = foo(q.to(torch.float64), k.to(torch.float64), v.to(torch.float64))
+    compiled_out = compiled(q, k, v)
+    torch.testing.assert_close(ref_out.to(dtype=torch.float16), compiled_out, atol=1e-1, rtol=0)
+    bench(lambda: foo(q, k, v), "eager")
+    bench(lambda: compiled(q, k, v), "compiled")
+
```

Comment:
Lol this is going to run when we import the file. You probably want to add this as a test somewhere? Or guard it using `if __name__ == '__main__':`